prov/util/src/util_ep.c

```diff
@@ -94,6 +94,84 @@ int ofi_ep_bind_av(struct util_ep *util_ep, struct util_av *av)
 	return 0;
 }
 
+int ofi_ep_bind_cntr(struct util_ep *ep, struct util_cntr *cntr, uint64_t flags)
+{
+	int ret;
+
+	ret = ofi_check_bind_cntr_flags(ep, cntr, flags);
+	if (ret)
+		return ret;
+
+	if (flags & FI_TRANSMIT) {
+		ep->tx_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & FI_RECV) {
+		ep->rx_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & FI_READ) {
+		ep->rd_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & FI_WRITE) {
+		ep->wr_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & FI_REMOTE_READ) {
+		ep->rr_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & FI_REMOTE_WRITE) {
+		ep->rw_cntr = cntr;
+		ofi_atomic_inc32(&cntr->ref);
+	}
+
+	if (flags & (FI_TRANSMIT | FI_RECV | FI_READ | FI_WRITE |
+		     FI_REMOTE_READ | FI_REMOTE_WRITE)) {
+		return fid_list_insert(&cntr->ep_list,
+				       &cntr->ep_list_lock,
+				       &ep->ep_fid.fid);
+	}
```

Comment:
right, fid_list_insert checks for existing entry. the same for removing: it will be removed only once